### PR TITLE
Add cargo-flux binary

### PR DESCRIPTION
**Pull request by cole-k**
# Edit: Usage and installation details are now in `flux-docs`

# Testing locally

Change https://github.com/liquid-rust/flux/pull/305/files#diff-e3670cdab561aaacb93c3521ffc0db64aa666949dcfc0e5e7f0148fd595ef5afR18 to your flux binary path.

```
cargo install --path flux/bin
```

Note that this will clobber a pre-existing binary, so if you put your `cargo-flux` shell script there, you'll need to copy it elsewhere if you don't want it to disappear.

### CODE REVIEW

**Comment by ranjitjhala** [comment]
Nice! Can you update the relevant bit in the docs?

**Comment by cole-k** [comment]
Haven't updated the docs yet but you should now be able to just do

```
cargo install --lib flux
cargo install --lib flux-bin
```

I'd love to just be able to do `cargo install`, but seems [that isn't possible](https://github.com/rust-lang/cargo/issues/4101) with the way things are set up.

@nilehmann any suggestions on how we might be able to reconfigure? Should I add a `[package]` and containing `[[bin]]`s to the main `Cargo.toml`?

**Comment by cole-k** [comment]
We could also add a binary that generates the vscode plugin json.

```json
{
  "rust-analyzer.checkOnSave.extraEnv": {
    "RUSTC_WRAPPER": "/path/to/flux/target/release/flux",
    "RUSTUP_TOOLCHAIN": "nightly-2022-10-11",
    "LD_LIBRARY_PATH": "/path/to/.rustup/toolchains/nightly-2022-10-11-x86_64-apple-darwin/lib"
  }
}
```

Although personally I feel like it would make more sense to have this be generated when you run `cargo build`. Not sure how to do that or where to put it, though.

**Comment by nilehmann** [comment]
I think this is good enough. You can do `cargo build --release` that will put all the binaries in `/path/to/flux/target/release` and then you can set up your vscode as

```
{
  "rust-analyzer.checkOnSave.overrideCommand": [
    "/path/to/flux/target/release/cargo-flux",
    "check",
    "--workspace",
    "--message-format=json"
  ]
}
```

File: flux-bin/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+pub const LIB_PATH: &str = "PATH";
+#[cfg(target_os = "linux")]
+pub const LIB_PATH: &str = "LD_LIBRARY_PATH";
+#[cfg(target_os = "macos")]
+pub const LIB_PATH: &str = "DYLD_FALLBACK_LIBRARY_PATH";
+
+pub const EXIT_ERR: i32 = -1;
+
+// It's better than using .expect(), right?
+pub fn report_err(message: &str, err: impl Error) -> i32 {
```
Comment (nilehmann) [inline flux-bin/src/utils.rs:13]: Maybe better to just return `Result<T, Box<dyn std::error::Error>>` everywhere and then print at the top level. The `description` will probably be good enough (and maybe better than our custom messages)  so no point in categorizing all the errors.

Comment (cole-k) [reply flux-bin/src/utils.rs:13]: Yeah I think that makes sense. I'm rendering the `impl Error` in addition to outputting a message, but looking through the file, it seems like there isn't much helpful.

**Comment by cole-k** [comment]
> I think this is good enough. You can do `cargo build --release` that will put all the binaries in `/path/to/flux/target/release` and then you can set up your vscode as
> 
> ```json
> {
>   "rust-analyzer.checkOnSave.overrideCommand": [
>     "/path/to/flux/target/release/cargo-flux",
>     "check",
>     "--workspace",
>     "--message-format=json"
>   ]
> }
> ```

If you ran `cargo install --lib flux/bin` would you be able to even write it as below?

```json
{
  "rust-analyzer.checkOnSave.overrideCommand": [
    "cargo-flux",
    "check",
    "--workspace",
    "--message-format=json"
  ]
}
```

File: flux-bin/src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    toolchain_file
+        .toolchain()
+        .spec()
+        .ok_or_else(|| report_message("No spec in rust-toolchain file"))?
+        .channel()
+        .ok_or_else(|| report_message("No channel in rust-toolchain file"))
+        .map(|channel| channel.name().to_string())
+}
+
+pub fn get_dyld_fallback_library_path(rust_toolchain: &str) -> Result<PathBuf, i32> {
```
[the remark this answers is not in the document]
Comment (cole-k) [reply flux-bin/src/utils.rs:54]: I picked the linux name (`ld_library_path`) since it seems the most descriptive, but I'm open to suggestions.

**Comment by nilehmann** [comment]
If  `cargo-flux` is in your path and `flux` is in the same directory then yes you would be able to write it like that. 

But what is this `cargo install --lib`? It doesn't work for me. I can do `cargo install --path flux-bin` and `cargo install --path flux`.

**Comment by cole-k** [comment]
> If `cargo-flux` is in your path and `flux` is in the same directory then yes you would be able to write it like that.
> 
> But what is this `cargo install --lib`? It doesn't work for me. I can do `cargo install --path flux-bin` and `cargo install --path flux`.

Yeah sorry, I meant `cargo install --path`. Fixed.

File: flux-bin/src/utils.rs (hunk 1 above)
Comment (cole-k) [reply flux-bin/src/utils.rs]: We now use `anyhow` to collect the errors.

File: flux-bin/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                    .map_err(anyhow::Error::from);
+            }
+        }
+    }
+    Err(anyhow!("Could not read Rustup toolchains folder"))
+}
+
+pub fn get_rustup_home() -> Result<PathBuf> {
+    env::var("RUSTUP_HOME")
+        .map(|home_dir| PathBuf::from(&home_dir))
```
Comment (nilehmann) [inline flux-bin/src/utils.rs:71]:
```suggestion
        .map(PathBuf::from)
```

File: flux-bin/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+pub const LIB_PATH: &str = "PATH";
+#[cfg(target_os = "linux")]
+pub const LIB_PATH: &str = "LD_LIBRARY_PATH";
+#[cfg(target_os = "macos")]
+pub const LIB_PATH: &str = "DYLD_FALLBACK_LIBRARY_PATH";
+
+pub const EXIT_ERR: i32 = -1;
+
+pub fn get_flux_path() -> Result<PathBuf> {
+    let mut flux_path = env::current_exe().map(|path| path.with_file_name("flux"))?;
```
Comment (nilehmann) [inline flux-bin/src/utils.rs]:
```suggestion
    let mut flux_path = env::current_exe()?.with_file_name("flux"));
```

File: flux-bin/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                }
+                _ => Err(anyhow::Error::from(e)),
+            }
+        })
+}
+
+/// Prepends the path so that it's the first checked.
+pub fn extend_env_var_with_path(var_name: &str, new_path: PathBuf) -> Result<OsString> {
+    let mut paths = env::var(var_name)
+        .map(|paths_str| env::split_paths(&paths_str).collect::<Vec<_>>())
```
Comment (nilehmann) [inline flux-bin/src/utils.rs:87]: I prefer `collect_vec`, but you would need to add `itertools` as a dependency.

File: flux/Cargo.toml [hunk 6]
```diff
@@ -2,6 +2,13 @@
 edition = "2021"
 name = "flux"
 version = "0.1.0"
+default-run = "flux"
```
Comment (nilehmann) [inline flux/Cargo.toml:5]: 👍🏽

**Comment by cole-k** [comment]
I've updated the readme with the new instructions. I tried to follow them myself but it would help if one of you could confirm that it works.

I also added a `FLUX_PATH` env variable so you can run a different version of flux if desired. I feel like there isn't really a difference between running

```bash
cargo install --lib flux
```

and


```bash
cargo build
```

but maybe if you have the muscle memory for the latter, you will find that env variable helpful.

**Review by nilehmann — COMMENTED** [review]
@cole-k the code looks good. Just left you some small stylistics suggestions. 

@ranjitjhala made this page https://liquid-rust.github.io/flux/. We you should add the changes to the docs there. The page can be edited here https://github.com/liquid-rust/flux/tree/main/flux-docs/src. We should also probably remove most of the text from the README and just put links to the doc

File: flux-bin/src/utils.rs (hunk 5 above)
Comment (cole-k) [reply flux-bin/src/utils.rs:87]: The annotation isn't even necessary, I think my intellisense inserted it before it knew the whole type.

**Comment by cole-k** [comment]
Moved the readme's contents to `flux-docs`.

File: flux-docs/src/use/install.md [hunk 7]
```diff
@@ -23,8 +23,17 @@ To build the source you need a nightly version of rustc.
...
+```
+
+Note that you should not call `flux` directly.
+
+Finally, install `rustc-flux` and `cargo-flux` using
+
+```bash
+cargo install --path flux-bin
+```
+```
```
Comment (ranjitjhala) [inline flux-docs/src/use/install.md:39]: extra ``` ?

Comment (cole-k) [reply flux-docs/src/use/install.md]: Thanks for that catch

**Comment by nilehmann** [comment]
@cole-k I think this is ready to go as soon as you resolve the conflict

**Review by nilehmann — APPROVED** [review]

**Pull request merged**